exonum-node/src/events/network.rs

```diff
@@ -261,245 +229,207 @@ impl NetworkHandler {
-            })
+                if connect_list.is_peer_allowed(&peer_key) {
+                    let conn_addr = ConnectedPeerAddr::In(peer_address);
+                    let receiver_rx = pool.add_incoming_address(peer_key, conn_addr.clone());
+                    let socket = Framed::new(socket, codec);
+                    let connection = Connection::new(socket, receiver_rx, conn_addr, peer_key);
+                    Self::handle_connection(connection, connect, pool, network_tx).await?;
+                } else {
+                    bail!(
+                        "Rejecting incoming connection with peer={} public_key={}, \
+                         the peer is not in the connect list",
+                        peer_address,
+                        peer_key
+                    );
+                }
+
+                drop(holder);
+                Ok(())
+            };
+
+            tokio::spawn(task.unwrap_or_else(|err| log::warn!("{}", err)));
+        }
+        Ok(())
     }
 
     fn connect(
         &self,
         key: PublicKey,
         handshake_params: &HandshakeParams,
-    ) -> impl Future<Item = (), Error = failure::Error> + Send {
-        let handshake_params = handshake_params.clone();
-        let handle = self.handle.clone();
-        let network_tx = self.network_tx.clone();
-        let network_config = self.network_config;
-        let timeout = self.network_config.tcp_connect_retry_timeout;
-        let max_tries = self.network_config.tcp_connect_max_retries as usize;
-        let max_connections = self.network_config.max_outgoing_connections;
-        let strategy = FixedInterval::from_millis(timeout)
-            .map(jitter)
-            .take(max_tries);
+    ) -> impl Future<Output = Result<(), failure::Error>> {
+        // Resolve peer key to an address.
+        let maybe_address = self.connect_list.find_address_by_key(&key);
+        let unresolved_address = if let Some(address) = maybe_address {
+            address
+        } else {
+            let err = format_err!("Trying to connect to peer not from ConnectList key={}", key);
+            return future::err(err).left_future();
+        };
 
-        let unresolved_address = self.connect_list.find_address_by_key(&key);
+        let max_connections = self.network_config.max_outgoing_connections;
+        let (sender_tx, receiver_rx) = mpsc::channel(OUTGOING_CHANNEL_SIZE);
+        let network_config = self.network_config;
+        let mut handshake_params = handshake_params.clone();
+        handshake_params.set_remote_key(key);
+        let pool = self.pool.clone();
+        let network_tx = self.network_tx.clone();
 
-        if let Some(unresolved_address) = unresolved_address {
-            let action = {
-                let unresolved_address = unresolved_address.clone();
-                move || tokio_dns::TcpStream::connect(unresolved_address.as_str())
+        // TODO: use retries.
```

Comment:
Do you plan to add retries in this PR or later?

Reply:
In this PR.